File: image_generation.js
const { createCanvas, loadImage, registerFont } = require('canvas');

async function generateImage(imagePath, query, title, subtitle) {
    const canvas = createCanvas(1000, 1000);
    const ctx = canvas.getContext('2d');

    const image = await loadImage(imagePath);
    const width = image.width;
    const height = image.height;

    ctx.drawImage(image, 0, 0, width, height, 0, 0, 1000, 1000);

    const imageData = ctx.getImageData(0, 0, 1000, 1000);
    const pixelData = imageData.data;

    let rSum = 0;
    let gSum = 0;
    let bSum = 0;
    let count = 0;

    for (let i = 0; i < pixelData.length; i += 4) {
        const r = pixelData[i];
        const g = pixelData[i + 1];
        const b = pixelData[i + 2];
        const a = pixelData[i + 3];

        rSum += r;
        gSum += g;
        bSum += b;
        count++;
    }

    const averageR = Math.floor(rSum / count);
    const averageG = Math.floor(gSum / count);
    const averageB = Math.floor(bSum / count);

    ctx.fillStyle = `rgba(${averageR}, ${averageG}, ${averageB}, 1)`;
    ctx.fillRect(0, 0, 1000, 1000);

    ctx.beginPath();
    ctx.arc(500, 500, 256, 0, Math.PI * 2);
    ctx.fillStyle = 'rgba(255, 255, 255, 1)';
    ctx.fill();
    ctx.lineWidth = 10;
    ctx.strokeStyle = 'rgba(255, 255, 255, 1)';
    ctx.stroke();

    const maxSize = Math.max(width, height);
    const resizeFactor = 512 / maxSize;
    const resizedWidth = width * resizeFactor;
    const resizedHeight = height * resizeFactor;

    ctx.drawImage(image, 0, 0, width, height, (1000 - resizedWidth) / 2, (1000 - resizedHeight) / 2, resizedWidth, resizedHeight);

    ctx.font = '80px "Arial"';
    ctx.fillStyle = 'white';
    const text1Width = ctx.measureText(title).width;
    ctx.fillText(title, (1000 - text1Width) / 2, (1000 + resizedHeight) / 2 + 50);

    ctx.font = '45px "Arial"';
    const text2Width = ctx.measureText(subtitle).width;
    ctx.fillText(subtitle, (1000 - text2Width) / 2, (1000 + resizedHeight) / 2 + 50 + 80);

    const fs = require('fs');
    const out = fs.createWriteStream(`assets/second/${query}.png`);
    const stream = canvas.createPNGStream();
    stream.pipe(out);
}

generateImage('image.png', 'query', 'Title', 'Subtitle');

File: image_generation.java
import javax.media.jai.*; // надо установить Java Advanced Imaging
import java.awt.*;
import java.awt.image.*;
import java.io.File;
import java.io.IOException;

public class ImageGenerator {
    public static void generateImage(String imagePath, String query, String title, String subtitle) throws IOException {
        RenderedImage image = JAI.create("fileload", imagePath);

        int width = image.getWidth();
        int height = image.getHeight();

        double[][] pixelColors = new double[width][height];
        int numBands = image.getSampleModel().getNumBands();

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                double[] pixel = new double[numBands];
                image.getData().getPixel(x, y, pixel);
                pixelColors[x][y] = (pixel.length == 1) ? pixel[0] : (pixel[0] + pixel[1] + pixel[2]) / 3.0;
            }
        }

        double rSum = 0, gSum = 0, bSum = 0;
        int count = 0;

        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                double pixelValue = pixelColors[x][y];
                rSum += pixelValue;
                gSum += pixelValue;
                bSum += pixelValue;
                count++;
            }
        }

        int averageR = (int) (rSum / count);
        int averageG = (int) (gSum / count);
        int averageB = (int) (bSum / count);

        BufferedImage background = new BufferedImage(1000, 1000, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = background.createGraphics();
        g2d.setColor(new Color(averageR, averageG, averageB, 255));
        g2d.fillRect(0, 0, 1000, 1000);

        BufferedImage circleImage = new BufferedImage(512, 512, BufferedImage.TYPE_INT_ARGB);
        Graphics2D circleG2d = circleImage.createGraphics();
        circleG2d.setColor(new Color(255, 255, 255, 0));
        circleG2d.fillOval(0, 0, 511, 511);
        circleG2d.setColor(new Color(255, 255, 255, 255));
        circleG2d.setStroke(new BasicStroke(10));
        circleG2d.drawOval(0, 0, 511, 511);

        double maxSize = Math.max(width, height);
        double resizeFactor = 512.0 / maxSize;
        int resizedWidth = (int) (width * resizeFactor);
        int resizedHeight = (int) (height * resizeFactor);

        BufferedImage resizedImage = new BufferedImage(resizedWidth, resizedHeight, BufferedImage.TYPE_INT_ARGB);
        Graphics2D resizedG2d = resizedImage.createGraphics();
        resizedG2d.drawImage(image, 0, 0, resizedWidth, resizedHeight, null);

        int offsetX = (1000 - resizedWidth) / 2;
        int offsetY = (1000 - resizedHeight) / 2;

        background.createGraphics().drawImage(circleImage, offsetX, offsetY, null);
        background.createGraphics().drawImage(resizedImage, offsetX, offsetY, null);

        Font bigFont = new Font("Arial", Font.BOLD, 80);
        Font smallFont = new Font("Arial", Font.PLAIN, 45);

        g2d.setFont(bigFont);
        g2d.setColor(Color.WHITE);
        FontMetrics bigFontMetrics = g2d.getFontMetrics();
        int text1Width = bigFontMetrics.stringWidth(title);
        g2d.drawString(title, (1000 - text1Width) / 2, offsetY + resizedHeight + 50);

        g2d.setFont(smallFont);
        g2d.setColor(Color.WHITE);
        FontMetrics smallFontMetrics = g2d.getFontMetrics();
        int text2Width = smallFontMetrics.stringWidth(subtitle);
        g2d.drawString(subtitle, (1000 - text2Width) / 2, offsetY + resizedHeight + bigFontMetrics.getHeight() + 60);

        File output = new File("assets/second/" + query + ".png");
        try {
            ImageIO.write(background, "PNG", output);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) throws IOException {
        generateImage("image.png", "query", "Title", "Subtitle");
    }
}
